### PR TITLE
Sphinx documentation

### PR DESCRIPTION
**Pull request by Bobbins228**
# Issue link

Closes: [RHOAIENG-13962](https://issues.redhat.com/browse/RHOAIENG-13962) and [RHOAIENG-9511](https://issues.redhat.com/browse/RHOAIENG-9511)
# What changes have been made

includes changes from #698 
* Added Sphinx code documentation automation Work Flow
* Added new documentation under Ray Cluster Interaction, UI Widgets and Cluster Configuration
* Proven working [workflow](https://github.com/Bobbins228/codeflare-sdk/actions/runs/11343674817/job/31546740644)
* Example of what the documentation will look like when the WF is run on main https://bobbins228.github.io/codeflare-sdk/
# Verification steps

N/A 
# TODO
Change the GitHub Pages deployment to use the new gh-pages branch from root dir.
Run the Workflow
## Checks
- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] Testing is not required for this change

### CODE REVIEW

File: .github/workflows/publish-documentation.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+          poetry install --with docs
+          sed -i 's/release = "v[0-9]\+\.[0-9]\+\.[0-9]\+"/release = "${{ github.event.inputs.codeflare_sdk_release_version }}"/' docs/sphinx/conf.py
+          sphinx-apidoc -o docs/sphinx src/codeflare_sdk "**/*test_*" --force # Generate docs but ignore test files
+          make html -C docs/sphinx
+      - name: Deploy to GitHub Pages
+        uses: peaceiris/actions-gh-pages@v3
+        with:
+          publish_branch: gh-pages
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          publish_dir: docs/sphinx/_build/html
```
Comment (Ygnas) [inline .github/workflows/publish-documentation.yaml:44]: Can we add this to .gitignore?

File: docs/sphinx/user-docs/authentication.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+Method 2 Kubernetes Config File Authentication (Default location)
+-----------------------------------------------------------------
+
+If a user has authenticated to their cluster by alternate means e.g. run
+a login command like ``oc login --token=<token> --server=<server>``
+their kubernetes config file should have updated. If the user has not
+specifically authenticated through the SDK by other means such as
+``TokenAuthentication`` then the SDK will try to use their default
+Kubernetes config file located at ``"/HOME/.kube/config"``.
```
Comment (Ygnas) [inline docs/sphinx/user-docs/authentication.rst:42]:
```suggestion
Kubernetes config file located at ``"$HOME/.kube/config"``.
```

File: docs/sphinx/user-docs/authentication.rst [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+       kube_config_path="/path/to/config",
+   )
+   auth.load_kube_config()
+   # log out with auth.logout()
+
+Method 4 In-Cluster Authentication
+----------------------------------
+
+If a user does not authenticate by any of the means detailed above and
+does not have a config file at ``"/HOME/.kube/config"`` the SDK will try
```
Comment (Ygnas) [inline docs/sphinx/user-docs/authentication.rst:65]:
```suggestion
does not have a config file at ``"$HOME/.kube/config"`` the SDK will try
```

File: docs/sphinx/user-docs/s3-compatible-storage.rst [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+   trainer = TorchTrainer(
+       train_func_distributed,
+       scaling_config=scaling_config,
+       run_config = get_minio_run_config()
+   )
+
+To find more information on creating a Minio Bucket compatible with
+RHOAI you can refer to this
+`documentation <https://ai-on-openshift.io/tools-and-applications/minio/minio/>`__.
+Note: You must have ``sf3s`` and ``pyarrow`` installed in your
```
Comment (Ygnas) [inline docs/sphinx/user-docs/s3-compatible-storage.rst:85]:
```suggestion
Note: You must have ``s3fs`` and ``pyarrow`` installed in your
```

File: pyproject.toml [hunk 5]
```diff
@@ -34,7 +34,8 @@ ipywidgets = "8.1.2"
 optional = true
 
 [tool.poetry.group.docs.dependencies]
-pdoc3 = "0.10.0"
+sphinx = "7.4.7"
```
Comment (Ygnas) [inline pyproject.toml:37]: Should we use newer sphinx?

File: .github/workflows/publish-documentation.yaml (hunk 1 above)
Comment (Bobbins228) [reply .github/workflows/publish-documentation.yaml:44]: Need to double check that this wont affect the WF

File: pyproject.toml (hunk 5 above)
Comment (Bobbins228) [reply pyproject.toml:37]: `7.4.7` is the latest version of Sphinx that is still supported in Python 3.9

File: .github/workflows/release.yaml [hunk 6]
```diff
@@ -51,33 +49,17 @@ jobs:
             - name: Run poetry install
               run: poetry install --with docs
             - name: Create new documentation
-              run: poetry run pdoc --html -o docs/detailed-documentation src/codeflare_sdk && pushd docs/detailed-documentation && rm -rf cluster job utils && mv codeflare_sdk/* . && rm -rf codeflare_sdk && popd && find docs/detailed-documentation -type f -name "*.html" -exec bash -c "echo '' >> {}" \;
+              run: |
```
[the remark this answers is not in the document]
Comment (Bobbins228) [reply .github/workflows/release.yaml:52]: Oh thats a good idea I will move it around

File: docs/sphinx/user-docs/cluster-configuration.rst [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+     - A dictionary of custom resource mappings to map extended resource requests to RayCluster resource names
+     - ``{}``
+   * - ``overwrite_default_resource_mapping``
+     - ``bool``
+     - A boolean indicating whether to overwrite the default resource mapping
+     - ``False``
+   * - ``local_queue``
+     - ``Optional[str]``
+     - A parameter for specifying the Local Queue label for the Ray Cluster
+     - ``None``
```
[the remark this answers is not in the document]
Comment (Bobbins228) [reply docs/sphinx/user-docs/cluster-configuration.rst:152]: No this is by hand but I can see this being removed as we have code comments in `config.py` which essentially say the same thing

File: .github/workflows/release.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Bobbins228) [reply .github/workflows/release.yaml:52]: Tried this out but it seems you cannot run an action within an action so the action would not be able to push the docs to the gh-pages branch.

File: docs/sphinx/user-docs/ui-widgets.rst [hunk 8]
```diff
@@ -0,0 +1,53 @@
+Jupyter UI Widgets
+==================
+
+Below are some examples of the Jupyter UI Widgets that are included in
+the CodeFlare SDK. > [!NOTE] > To use the widgets functionality you must
```
[the remark this answers is not in the document]
Comment (Bobbins228) [reply docs/sphinx/user-docs/ui-widgets.rst:5]: Great catch thanks Ignas.

**Review by Ygnas — APPROVED** [review]

**Review by KPostOffice — APPROVED** [review]

**Pull request merged**